mcs/class/System/System.Net/HttpWebRequest.cs

```diff
@@ -926,30 +945,34 @@ public Stream GetRequestStream (out TransportContext context)
 			throw new NotImplementedException ();
 		}
 
-		bool CheckIfForceWrite (SimpleAsyncResult result)
+		internal static async Task<T> RunWithTimeout<T> (Func<Task, CancellationToken, Task<T>> func, int timeout, Action abort)
 		{
-			if (writeStream == null || writeStream.RequestWritten || !InternalAllowBuffering)
-				return false;
-			if (contentLength < 0 && writeStream.CanWrite == true && writeStream.WriteBufferLength < 0)
-				return false;
-
-			if (contentLength < 0 && writeStream.WriteBufferLength >= 0)
-				InternalContentLength = writeStream.WriteBufferLength;
-
-			// This will write the POST/PUT if the write stream already has the expected
-			// amount of bytes in it (ContentLength) (bug #77753) or if the write stream
-			// contains data and it has been closed already (xamarin bug #1512).
+			using (var cts = new CancellationTokenSource ()) {
```

Comment:
Why do you need all this when you could do `task.Wait (timeout)` ?

Reply:
We need two things:
a) A `CancellationToken` that's automatically being cancelled after the timeout (with the `abort` delegate being called).
b) Return a Task that is guaranteed to complete within the timeout.

So using `task.Wait (timeout)` here doesn't help us here.

However, we don't have to actually pass the `timeoutTask` to the inner method, I'm going to fix that.

Reply:
Cleaned this up, is it better now?

Reply:
there is following override for exactly that
```
public bool Wait(
	int millisecondsTimeout,
	CancellationToken cancellationToken
)
```

Reply:
I still don't understand why you need to set up 2 identical cancellations. You setup timeoutthe cancellation timeout which will throw (and you are missing catch for this) and extra wait on timeout.

I think simple 
```
try {
await func (timeoutTask, cts.Token).ConfigureAwait (false);
} catch (OperationCanceledException) {
}
```
should work or if you think the cancellation is not imminent, don't set it with timeout and Cancell it manually when you get timeoutTask result.